libvast/src/segment.cpp

```diff
@@ -0,0 +1,201 @@
+  auto payload = chunk_->data() + header_.payload_offset;
+  // Walk in lock-step through the slices and the ID sequence.
+  for (auto& slice : meta_.slices) {
+    if (!rng)
+      return result;
+    // Make the ID range catch up if it's behind.
+    if (rng.get() < slice.offset) {
+      rng.skip(slice.offset);
+      if (!rng)
+        return result;
+    }
+    // If the next ID falls in the current slice, we add to the result.
+    if (rng.get() >= slice.offset && rng.get() <= slice.offset + slice.size) {
+      auto slice_size = detail::narrow_cast<size_t>(slice.end - slice.start);
+      caf::charbuf buf{payload + slice.start, slice_size};
+      caf::stream_deserializer<caf::charbuf&> deserializer{actor_system_, buf};
+      const_table_slice_handle x;
+      if (auto error = deserializer(x))
+        return error;
+      result.push_back(x);
+      // Fast forward to the ID one past this slice.
+      rng.skip(slice.offset + slice.size);
+    }
+  }
+  return result;
+}
+
+segment::segment(caf::actor_system& sys, chunk_ptr chunk) 
+  : actor_system_{sys},
+    chunk_{std::move(chunk)} {
+  // Only the builder and make() call this constructor. In the former case the
+  // chunk is properly constructed and in the latter case the factory ensures
+  // that it meets the requirements.
+  VAST_ASSERT(chunk_ != nullptr);
+}
+
+segment_builder::segment_builder(caf::actor_system& sys)
+  : actor_system_{sys},
+    table_slice_streambuf_{table_slice_buffer_},
+    table_slice_serializer_{actor_system_, table_slice_streambuf_} {
+  reset();
+}
+
+caf::error segment_builder::add(const_table_slice_handle x) {
+  if (x->offset() < min_table_slice_offset_)
+    return make_error(ec::unspecified, "table slice offsets not increasing");
+  auto before = table_slice_buffer_.size();
+  if (auto error = table_slice_serializer_(x)) {
+    table_slice_buffer_.resize(before);
+    return error;
+  }
+  auto after = table_slice_buffer_.size();
+  VAST_ASSERT(before < after);
+  meta_.slices.push_back({
+    detail::narrow_cast<int64_t>(before),
+    detail::narrow_cast<int64_t>(after),
+    x->offset(), x->rows()});
+  min_table_slice_offset_ = x->offset() + x->rows();
+  return caf::none;
+}
+
+caf::expected<segment_ptr> segment_builder::finish() {
+  auto guard = caf::detail::make_scope_guard([&] { reset(); });
+  // Write header.
+  segment_buffer_.resize(sizeof(segment::header));
+  auto header = reinterpret_cast<segment::header*>(segment_buffer_.data());
+  header->magic = to_little_endian(segment::magic);
+  header->version = to_little_endian(segment::version);
+  header->id = uuid::random();
+  // Serialize meta data into buffer.
+  caf::vectorbuf segment_streambuf{segment_buffer_};
+  detail::coded_serializer<caf::vectorbuf&> meta_serializer{segment_streambuf};
+  if (auto error = meta_serializer(meta_))
+    return error;
+  // Add the payload offset to the header.
+  header = reinterpret_cast<segment::header*>(segment_buffer_.data());
```

Comment:
Why is this assignment repeated?

Reply:
Because the vector grew in the meantime. Without this "reset", we'd have heap-use-after-free errors.